pkg/api/types.go

```diff
@@ -180,6 +180,113 @@ type VolumeSource struct {
+
+	// TODO (per thockin) https://github.com/GoogleCloudPlatform/kubernetes/pull/4055#discussion_r24865963
+	AWSElasticBlockStore *AWSElasticBlockStore `json:"elasticBlockStore"`
+
+	NFSMount *NFSMount `json:"nfsMount"`
+
+	PersistentVolumeClaimVolumeSource *PersistentVolumeClaimVolumeSource `json:"persistentVolumeClaim,omitempty"`
+}
+
+type PersistentVolume struct {
+	TypeMeta   `json:",inline"`
+	ObjectMeta `json:"metadata,omitempty"`
+
+	//Spec defines a persistent volume owned by the cluster
+	Spec PersistentVolumeSpec `json:"spec,omitempty"`
+
+	// Status represents the current information about persistent volume.
+	Status PersistentVolumeStatus `json:"status,omitempty"`
+
+	// holds the binding reference to a PersistentVolumeClaim
+	ClaimRef *ObjectReference `json:"claimRef,omitempty"`
+}
+
+type PersistentVolumeSpec struct {
+	Capacity ResourceList `json:"capacity,omitempty`
+	// AccessModeTypes are inferred from the Source
+	Source VolumeSource `json:",inline"`
+}
+
+type PersistentVolumeStatus struct {
+	Phase StoragePhase `json:"phase,omitempty"`
+}
+
+type PersistentVolumeList struct {
+	TypeMeta `json:",inline"`
+	ListMeta `json:"metadata,omitempty"`
+	Items    []PersistentVolume `json:"items,omitempty"`
+}
+
+// a PersistentVolumeClaim is a user's request for and claim to a persistent volume
+type PersistentVolumeClaim struct {
+	TypeMeta   `json:",inline"`
+	ObjectMeta `json:"metadata,omitempty"`
+
+	// Spec defines the volume
+	Spec PersistentVolumeClaimSpec `json:"spec,omitempty"`
+
+	// Status represents the current information about a persistent volume.
+	// this data may not be up to date.
+	Status PersistentVolumeClaimStatus `json:"status,omitempty"`
+}
+
+type PersistentVolumeClaimList struct {
+	TypeMeta `json:",inline"`
+	ListMeta `json:"metadata,omitempty"`
+	Items    []PersistentVolumeClaim `json:"items,omitempty"`
+}
+
+// a PersistentVolumeClaimSpec describes the common attributes of storage devices
+// and allows a Source for provider-specific attributes
+type PersistentVolumeClaimSpec struct {
+	// Contains the types of access modes desired
+	AccessModes []AccessModeType     `json:"accessModes,omitempty"`
+	Resources   ResourceRequirements `json:"resources,omitempty"`
+}
+
+type PersistentVolumeClaimStatus struct {
+	Phase       StoragePhase     `json:"phase,omitempty"`
+	AccessModes []AccessModeType `json:"accessModes,omitempty`
+	Requests    ResourceList     `json:"requests,omitempty"`
+	VolumeRef   *ObjectReference `json:"volumeRef,omitempty"`
+}
+
+type AccessModeType string
+
+const (
+	ReadWriteOnce AccessModeType = "ReadWriteOnce"
+	ReadOnlyMany  AccessModeType = "ReadOnlyMany"
+	ReadWriteMany AccessModeType = "ReadWriteMany"
+)
+
+type StoragePhase string
+
+const (
+	StoragePending StoragePhase = "Pending"
+	StorageBound   StoragePhase = "Bound"
+)
+
+type PersistentVolumeClaimVolumeSource struct {
+	AccessMode               AccessModeType   `json:"accessMode,omitempty"`
```

Comment:
This is a struct that users will specify - let's make it as simple as possible and document every field.

Why does the user need to spec AccessMode?  Presumably it's so that the IaaS knows that a multi-reader mount is OK vs a single-writer?  Document it.

Do we need a full ObjectReference?  That object itself is under-specced.  Can a user reference a claim in a different namespace than their own?  Do they need to know the name and UID or just name?  In other words, could we get away with just "ClaimName string"  here?

@bgrant0607